src/kern/exec.c

```diff
@@ -16,44 +16,51 @@
 #include "vmlinux.h"
 #include <bpf/bpf_helpers.h>
 #include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+#include <linux/limits.h>
 
-/*
- * Define a map structure to communicate with the userspace program.
- * The type of map is BPF_MAP_TYPE_PERCPU_ARRAY;
- * it contains one element, found for key = 0
- * and the value is an array with number_of_CPUs elements.
- */
-struct bpf_map_def SEC("maps") output_map = {
-	.type = BPF_MAP_TYPE_PERCPU_ARRAY,
-	.key_size = sizeof(int),
-	.value_size = sizeof(int),
-	.max_entries = 1,
+struct record_sample {
+	int ppid;
+	int pid;
+	int tgid;
+	char name[PATH_MAX];
 };
 
+struct {
+	__uint(type, BPF_MAP_TYPE_RINGBUF);
+	__uint(max_entries, 1 << 24);
+} ringbuf SEC(".maps");
+
+long ringbuf_flags = 0;
+
 SEC("lsm/bprm_committed_creds")
-int BPF_PROG(test_void_hook, struct linux_binprm *bprm)
+void BPF_PROG(exec_audit, struct linux_binprm *bprm)
 {
-	char str[] = "hello\n";
-	int key = 0; // the only key in the map, since there is just one key-value pair
-
-	// print a string for each executed process
-	bpf_trace_printk(str, sizeof(str));
-
-	/*
-	 * Obtain the current value found at key = 0 and on the current CPU.
-	 * val points to output_map[key][i], where i is the index of the current
-	 * CPU; this index is already known by the BPF program, as it internally
-	 * keeps track of the CPUs.
-	 */
-	int *val = bpf_map_lookup_elem(&output_map, &key);
-	if (!val) {
-		return 0;
-	}
-
-	// Increment the previous value, as a new process was executed on the
-	// current CPU.
-	(*val)++;
-	return 0;
+	int pid, tgid, ppid;
+	struct record_sample *sample;
+	struct task_struct *current_task;
+
+	// Get information about the current proces
+	pid = bpf_get_current_pid_tgid() >> 32;
+	tgid = (bpf_get_current_pid_tgid() << 32) >> 32;
+	ppid = -1;
+
+	sample = bpf_ringbuf_reserve(&ringbuf, sizeof(*sample), ringbuf_flags);
```

Comment:
What about trying to reserve space on the ringbuffer asap to bail out if this doesn't work (this is critical anyway) ? You could then directly write into sample->ppid, sample->pid etc... instead of having variables on the stack that you copy later